### ISSUE TITLE
Rule proposal: `prefer-array-from`

### ISSUE THREAD

**Issue by Pyrolistical** [sindresorhus/eslint-plugin-unicorn#1714]
### Description

Often [`no-for-loop`](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-for-loop.md) results in something that be simplified down to [`Array.from`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from)

Note for even simpler cases are handled by https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1713

This is an alternative to https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1712 and only one of these rules should be implemented.

Original idea: https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1712#issuecomment-1029310144

### Fail

const result = []
for (const element of array) {
  result.push(transform(element))
}
```
```js
const result = []
for (const [index, element] of array.entries()) {
  result[index] = transform(element)
}
```
```js
const result = []
for (const [index, element] of array.entries()) {
  result.push(anotherArray[index](element))
}
```

### Pass

```js
const result = Array.from(array, (element) => transform(element))
```
```js
const result = Array.from(array, (element, index) => anotherArray[index](element))
```

**Comment by Pyrolistical** [comment]
Another idea would be to roll this into `prefer-spread` in the form of

```js
const result = [...array].map((element) => transform(element))
```

**Comment by Pyrolistical** [comment]
This should handle `for await..of` in the future https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1712#issuecomment-1028915692

**Comment by fregante** [comment]
This is basically a "prefer array map" that also works for iterators. If the input is an actual array like in your example it's best to just just .map()

Also generally I'm not sure this is worth it for two reasons:

- there's already a rule the prefers [...it] over Array.from(it)
- there's not much difference between:
    ```JavaScript 
    Array.from(it, a => 2)
    [...it].map(a => 2)
    ```
    so it doesn't seem to be worth an exception from the rule above

**Comment by sindresorhus** [comment]
This is covered by #3220 now. The rule handles the core sync iterable cases from this issue by converting the empty-array + `for...of` accumulator pattern to `Array.from(iterable, mapper)`, including the `entries()` index cases.

I’m intentionally leaving `for await...of` out of scope here. That would need `Array.fromAsync()`, whose mapper awaits returned values, so it is not always equivalent to pushing the mapper result in the loop.
